Сделал блок 'hero' и 'marketplace'

Rubsee opened the pull request.

Rubsee committed:
Сделал блок 'hero' и 'marketplace'

captain-zsa commented on source/index.html:
```diff
@@ -63,6 +63,53 @@
+            <p class="hero__desc">Какой-то супер-мотивирующий текст<br>
+              чтоб купили прям вообще всё</p>
+          </div>
+          <button class="hero__button">
```
Таких кнопок уйма на сайте. Есть смысл сделать это бэм-блоком.

captain-zsa commented on source/index.html:
```diff
@@ -63,6 +63,53 @@
+              чтоб купили прям вообще всё</p>
+          </div>
+          <button class="hero__button">
+            <span>За покупками</span>
```
Сомневаюсь, что тут span вообще необходим.

captain-zsa commented on source/index.html:
```diff
@@ -63,6 +63,53 @@
+          </button>
+        </div>
+        <div class="hero__product">
+          <img class="hero__product--image" src="img/white_sneaker.png" alt="Белые кроссовки">
```
Это что такое? Вы делаете модификаторы почему-то. Принцип БЭМ – модификаторы МОДИФИЦИРУЮТ изначальное поведение. Изначальное поведение должно быть у бэм-блока или бэм-элемента. То есть

.button { всякие свойства типа padding, border и тп, а дальше идёт по-умолчанию для всех кнопок к примеру color: white; }
.button--green {color: green;} просто модифицирует конкретную кнопку, делая её цвет зеленым.

Это не отдельный элемент с таким классом модификатором и больше ничем. Это прям то, что надо модифицировать. Кнопка из примера выше будет такой примерно `<button class="button button--green">`

Более того, я давал для проверки нейминга по бэму сервис. Почему его не использовали? Или если и использовали – почему не спросили про ошибки, которые он высветил, если не поняли их? Он моментально подсветил эти ошибки. https://yoksel.github.io/html-tree/ вот ещё раз дублирую.

В общем переделывать надо.

captain-zsa commented on source/index.html:
```diff
@@ -63,6 +63,53 @@
+      </div>
+    </div>
+    <ul class="marketplace">
+      <li class="marketplace__ozon">
```
Это одинаковые элементы. marketplace__item для всех тегов li, а также если это вообще необходимо – добавляются модификаторы типа marketplace__item--ozon и так далее. Как раз за тем, чтобы сделать для всех li нужные одинаковые для всех свойства, а далее уже если и будет нужно что-то поменять точечно – изменить это применять свойства к модификатором.

captain-zsa commented on source/sass/blocks/container.scss:
```diff
@@ -1,6 +1,7 @@
 .container {
     margin: 0 auto;
-    max-width: 1560px;
+    max-width: 1728px;
```
Это почему же так? Весь контент находится как раз внутри контейнера 1560px а не внутри 1728px
<img width="967" alt="Снимок экрана 2024-06-18 в 11 26 38" src="https://github.com/Rubsee/gulp-start/assets/18168845/48e991b0-4648-4af1-8eac-2eae5c9bc32f">
вот для примера элемент и видно, что все остальные также расположены. У нас именно 1560 является переходом на другое разрешение (окей, да, чтоб было ещё по 10px в запасе чтоб не липло к краям экрана – то 1580 можно считать).

captain-zsa commented on source/sass/blocks/hero.scss:
```diff
@@ -0,0 +1,236 @@
+.hero {
+    $hr: '.hero';
```
Почему используется $bl в видео про вложенность – потому что так именуется переменная всегда и это удобно. Не надо запоминать, что где-то она $hr, где-то она как-то ещё называется и тп. Универсальное название $bl означает блок (от сокращения  $block чтоб просто длинное название не писать). Тогда в любом файле создаете (где нужно вообще это) такую переменную и никогда не будете думать "А вот у меня там файл marketplace.scss, у меня там название этой переменной, обозначающей блок, оно $mr сокращается или как-то иначе...". $bl универсальна всегда. Никаких конфликтов между файлами не будет, потому что эта переменная находится внтури селектора бэм-блока.

И да, писать '.hero' нет никакого смысла. Можно использовать $bl: &; и всё.

captain-zsa commented on source/sass/blocks/hero.scss:
```diff
@@ -0,0 +1,236 @@
+            margin-top: 226px;
+        }
+
+        #{$hr}__title {
```
Я чет совсем не понимаю, а для чего нужно делать эту дополнительную вложенность то? Зачем внутрь .hero__text нарочито вкладывать .hero__title чтоб получался селектор `.hero__text .hero__title` ? А ниже так вообще ужас просто. На выходе #{$hr}__heading span:first-child из вот этого кошмара получится

`.hero__text .hero__title .hero__heading span:first-child`

Специфичность этого селектора запредельна (он катастрофически сильный). Более того, рассчитать прям вот сходу, что это будет именно такой селектор, достаточно сложно. Запутаться можно в момент вообще, не уследив, что в чём там вложено.

Не надо так делать, если в этом нет действительно архи-важной нужды (такая нужда бывает крайне редко, к примеру когда у нас действительно внутри блока есть 2 элемента, один внутри другого, а другой нет и тогда да, нужно делать вложенность, но и то не такую большую, как тут).

В вашем случае такой нужды, делать такой огромный селеткор, действительно нет вообще. Это вложенность ради вложенности, усложнение ради усложнения, что только увеличивает количество кода в итоговом css, увеличивая время на обработку и скачивание этого файла.

captain-zsa commented on source/sass/blocks/hero.scss:
```diff
@@ -0,0 +1,236 @@
+                margin: 0;
+            }
+
+            #{$hr}__heading span:first-child {
```
:first-child и так далее крайне не удобны в использовании тут. А что если завтра попросят поменять шрифты? Верхний сделать такой, как нижний, а второй сделать такой как верхний? Придется сидеть и перелопачивать стили. А так удобно было бы иметь отдельные классы (возможно модификаторы) и не завязываться ещё и на конкретный тег span.

captain-zsa commented on source/sass/blocks/marketplace.scss:
```diff
@@ -0,0 +1,73 @@
+.marketplace {
+    @include clear-list;
+    display: flex;
+    height: 63px;
```
Не нужно тут задавать height. Вдруг завтра захотят делать второй ряд (еще один список добавят). И да, это прям отдельная секция, лучше так сделать. Чтоб было удобнее при необходимости ещё что-то в неё встроить. А внутри этой секции уже делать ul.

captain-zsa commented on source/sass/blocks/marketplace.scss:
```diff
@@ -0,0 +1,73 @@
+    &__ya {
+        img {
+            width: 90px;
+            height: 16px;
```
Тотально во всех картинках одинаковая высота (причем на планшете она тоже одинакова). Как раз этот код и можно было бы супер-сильно сократить за счет того, что вы сделали бы общий элемент marketplace__item и ему бы написали стили типа таких
```
&__item {
   img {
       height: 16px;
       ваш медиа с высотой
   }

   &--ozon {
        img {
            width: 73px;
            ваш медиа
        }
    }
}
```

и так со всеми собственно. А если стилей в img ещё куча добавится – будет ещё удобнее с таким подходом, потому что внутрь __item мы сможем встроить общее, а частное вынести в модификаторы.

Rubsee committed:
Работа над ошибками

Все поправил, но после смены max-width в центровшике, не могу нормально растянуть бэкграунд на планшетной версии. При ширине от 1560px и до 1728px если растягивать через (margin -20px) то все ровно не хватает, а если ставить примерно -100px чтоб зватило, тогда 'ultimate' уходит за границы.

captain-zsa commented on source/sass/blocks/preview-title.scss:
```diff
@@ -0,0 +1,53 @@
+        font-weight: 700;
+    }
+
+    &__item--your {
```
Зачем так делать? Можно же прямо внутрь &__item всё писать как 
```
&__item {
   ...свойства его...

  &--your {
       ....свойства...
   }
}
```

Rubsee committed:
Поправил недочеты

- центровщик
- вложенность в preview-title
